docs/terrain.md

```diff
@@ -0,0 +1,123 @@
+
+Voxel data is preceded by the offset between this chunk and the last in the blob, or from `0, 0, 0` if this is the first chunk in the blob. This offset is stored using 3 vectors of `(x: u8, y: u8, z: u8)` using the signedness determination described in [document conventions](#document-conventions), with 0xFF values in _all_ unused offsets indicating a negative sign. All coordinates are in chunk space (increments of 1 chunk), not world space.
+
+For example, given two chunks, one at a position of `(2, 0, 0)` and another at a position of `(4, 0, -1)` in chunk space, the latter's offset would be stored as follows:
+
+1. (0x00, 0x00, 0xFF)
+2. (0x00, 0x00, 0xFF)
+3. (0x00, 0x00, 0xFF)
+4. (0x02, 0x00, 0x01)
+
+| Field Name     | Format          | Value                                                                               |
+| :------------- | :-------------- | :---------------------------------------------------------------------------------- |
+| Signedness     | `[u8; 3]`       | Always the signedness of the following vectors. 0xFF in negative axes.              |
+| Offset (65536) | `[i8; 3]`       | Offset from the last chunk in the blob, each axis being multiplied by 65536 chunks. |
+| Offset (256)   | `[i8; 3]`       | Offset from the last chunk in the blob, each axis being multiplied by 256 chunks.   |
+| Offset (1)     | `[i8; 3]`       | Offset from the last chunk in the blob, each axis being in singular chunks.         |
+| Voxels         | [Voxel](#voxel) | As many voxels as needed to fulfill the maximum amount per chunk.                   |
+
+### Voxel
+
+The `Voxel` type is stored with a dynamic size (between 1-4 bytes) dependent on its set [bitflags](#voxel.flag). Only stored within [chunks](#chunk). A reference for the data contained within voxels [can be found here][Terrain.WriteVoxelChannels]. Empty voxels are stored with their material set to Air.
+
+Voxels are stored within a chunk using [run-length encoding]. A count of `1` would indicate the voxel repeating itself once, meaning 2 of the same voxel in a row would be read when decoded. Occupancy values are stored using the float format described in the [document conventions](#document-conventions).
+
+Water occupancy was added as an addendum to the existing format for the new [Shorelines] feature release. To avoid restricting the [Voxel.Flag](#voxel.flag)'s `Material Index` field to 5 bits, or rearchitecting the format, run-length encoding is disabled for the current voxel when it has a water occupancy of above `0.0`. Instead, the `Store Count` bit is enabled, and a count of `0` is written. Following the empty count, the water occupancy byte is stored. If a voxel has a solid occupancy of `1.0` (unless the material is Air), the water occupancy should always be `0.0`.
```

Comment:
This feels like it could be simplified somewhat. I would like to provide a suggestion, but it's unclear what it means to me. Under what conditions is water occupancy stored? My best guess is it's when `Store Count` is enabled and the count is written as `0`. If this is accurate, I'd write it like this:

```suggestion
Water occupancy is dynamically set based on the `Store Count` flag and the stored `Count` value. If `Store Count` is enabled and `Count` is `0x00`, then the Water Occupancy is written.

If the occupancy of a Voxel is `1.0` and the set material is not `Air`, the water occupancy must always be `0.0` if it's set.
```